### PR TITLE
Initial commit for chaindatafetcher

### PR DESCRIPTION
**Pull request by jeongkyun-oh**
## Proposed changes

- Implemented `ChainDataFetcher` structure which can be started with EN options.
- ChainDataFetcher can extract raw chain-data, transform it based on a specific database scheme, and load it on the database.
  - It can do the job in near-real-time.
  -  It can do the job based on API requests.
- This PR only writes logs to check if the components work well with the near-real-time job.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+	return nil
+}
+
+func (f *ChainDataFetcher) Stop() error {
+	f.scheduler.Stop()
+	for _, h := range f.handlers {
+		h.Stop()
+	}
+	logger.Info("chaindatafetcher is stopped")
+	return nil
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/chaindata_fetcher.go:62]: How about add a wait group to wait for scheduler and handlers to stop?

File: datasync/chaindatafetcher/scheduler.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (s *scheduler) reqLoop() {
+	for {
+		select {
+		case <-s.reqStopCh:
+			logger.Info("scheduler request loop is stopped")
+			return
+		case ev := <-s.chainCh:
+			logger.Info("scheduler: send", "blockNumber", ev.Block.Number().String())
```
Comment (KimKyungup) [inline datasync/chaindatafetcher/scheduler.go:74]:
```suggestion
			logger.Info("scheduler: send", "blockNumber", ev.Block.NumberU64())
```

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go:69]: Is a wait group necessary for all schedulers to be stopped? I think chaindata_handlers can be stopped in parallel. 6699a00f7a99a3e8a4102c62746154afda1381fd

File: datasync/chaindatafetcher/scheduler.go (hunk 2 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/scheduler.go]: Applied. 7983a09bd58b294c35e76522b0d03a3ddac10287

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	logger.Info("chaindatafetcher is started")
+	return nil
+}
+
+func (f *ChainDataFetcher) Stop() error {
+	f.scheduler.Stop()
+	wg := sync.WaitGroup{}
+	for _, h := range f.handlers {
+		wg.Add(1)
+		go func(handler ChainDataHandler) {
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:62]: Do we need to use goroutines to stop the handlers?

File: cmd/utils/flags.go [hunk 4]
```diff
@@ -591,6 +592,57 @@ var (
...
 	}
+	// ChainDataFetcher
+	EnableChainDataFetcherFlag = cli.BoolFlag{
+		Name:  "chaindatafetcher",
+		Usage: "Enable the ChainDataFetcher Service",
+	}
+	ChainDataFetcherNoDefault = cli.BoolFlag{
+		Name:  "chaindatafetcher.no.default",
+		Usage: "Turn off the starting of the chaindatafetcher",
+	}
```
Comment (ehnuje) [inline cmd/utils/flags.go:603]: How about merging these two flags?

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 1 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go:69]: Yes, it is  `Service` interface in enroll as a EN service.

File: datasync/chaindatafetcher/types.go [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package chaindatafetcher
+
+import "github.com/klaytn/klaytn/blockchain"
+
+type RequestType int
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/types.go]: Not yet, but will be used in the future. Removed it for now. 44c012b2d6ead818ee6181c595912ea989aad284

File: datasync/chaindatafetcher/scheduler.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (s *scheduler) resLoop() {
+	for {
+		select {
+		case <-s.resStopCh:
+			logger.Info("scheduler response loop is stopped")
+			return
+		case res := <-s.resCh:
+			logger.Info("scheduler: receive", "blockNumber", res.req.event.Block.Number().Uint64())
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/scheduler.go:87]: Hmm. Yes, `scheduler` will gather the results of requests and update the checkpoint.

File: datasync/chaindatafetcher/scheduler.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+func (s *scheduler) reqLoop() {
+	for {
+		select {
+		case <-s.reqStopCh:
+			logger.Info("scheduler request loop is stopped")
+			return
+		case ev := <-s.chainCh:
+			logger.Info("scheduler: send", "blockNumber", ev.Block.Number().Uint64())
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/scheduler.go]: Right. Applied on a7780751ffcfe02de93e22bb597d8e0d027dfec9

File: datasync/chaindatafetcher/scheduler.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+func (s *scheduler) StopWithRange(processId int) error {
+	panic("implement me")
+}
+
+func (s *scheduler) SetBlockchain(blockchain *blockchain.BlockChain) {
+	s.blockchain = blockchain
+}
+
+func (s *scheduler) reqLoop() {
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/scheduler.go:65]: If a loop is used, wouldn't it be possible to be hanged? 

scheduler --req--> [req, req, req, req, req] 
[res, res, res, res, res] <--res-- chaindata_handlers

File: datasync/chaindatafetcher/chaindata_handler.go [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+func (h *chainDataHandler) Stop() {
+	h.stopCh <- struct{}{}
+}
+
+func (h *chainDataHandler) Extract(req Request) interface{} {
+	logger.Info("handler: receive", "blockNumber", req.event.Block.Number().Uint64(), "id", h.id)
+	return req.event
+}
+
+func (h *chainDataHandler) Transform(i interface{}) interface{} {
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_handler.go]: Well, currently it is a bit vague. I will add them again when I implement them.

File: datasync/chaindatafetcher/chaindata_handler.go [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+			h.Load(i)
+		}
+	}
+}
+
+func (h *chainDataHandler) Stop() {
+	h.stopCh <- struct{}{}
+}
+
+func (h *chainDataHandler) Extract(req Request) interface{} {
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_handler.go]: I will add the comments when I implement them.

File: datasync/chaindatafetcher/chaindata_handler.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+func (h *chainDataHandler) Load(i interface{}) {
+	ev := i.(blockchain.ChainEvent)
+	h.resCh <- Response{
+		req: Request{
+			event: ev,
+		},
+		err: nil,
+	}
+	logger.Info("handler: send", "blockNumber", ev.Block.Number().Uint64(), "id", h.id)
```
[the remark this answers is not in the document]
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_handler.go]: Right. I updated the log messages.

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 3 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Nope. I modified logics to handle goroutines.

**Comment by jeongkyun-oh** [comment]
@ehnuje @KimKyungup Sorry to bother you, but I introduced chaindata handler manager to handle goroutines. Would you please take another look? Thanks in advance.

File: cmd/utils/flags.go (hunk 4 above)
Comment (jeongkyun-oh) [reply cmd/utils/flags.go:603]: Do you mean always auto-starting when `--chaindatafetcher` option is on?

Comment (ehnuje) [reply cmd/utils/flags.go:603]: Right, when do we need a chain data fetcher which is given both EnableChainDataFetcherFlag and ChainDataFetcherNoDefault ?

File: cmd/utils/flags.go [hunk 12]
```diff
@@ -591,6 +592,57 @@ var (
...
+	ChainDataFetcherReqChSize = cli.IntFlag{
+		Name:  "chaindatafetcher.request.buffer",
+		Usage: "Buffer size of request channel",
+		Value: 50,
+	}
+	ChainDataFetcherResChSize = cli.IntFlag{
+		Name:  "chaindatafetcher.response.buffer",
+		Usage: "Buffer size of response channel",
+		Value: 50,
+	}
```
Comment (aidan-kwon) [inline cmd/utils/flags.go:618]: How about using one variable for channels?

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+func (f *ChainDataFetcher) APIs() []rpc.API {
+	// TODO-ChainDataFetcher add APIs to start or stop chaindata fetcher
+	return []rpc.API{}
+}
+
+func (f *ChainDataFetcher) Start(server p2p.Server) error {
+	// launch multiple goroutines to handle new blocks
+	for i := 0; i < f.numHandlers; i++ {
+		f.wg.Add(1)
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:55]: This seems to be just a matter of taste, however, how about passing wait group as a parameter of handler and call `Add` and `Done` inside the handler? In that case, we can remove a wrapping function and just call `go f.handler(f.wg)`

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+		case <-f.stopCh:
+			logger.Info("stopped main loop for chaindatafetcher")
+			return
+		case ev := <-f.chainCh:
+			num := ev.Block.NumberU64()
+			f.reqCh <- num
+			logger.Info("request to handle new block", "blockNumber", num)
+		case res := <-f.resCh:
+			f.updateCheckpoint(res)
+			logger.Info("processed requested block", "blockNumber", res)
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:121]: Logs from `f.chainCh` and `f.resCh` could be noisy.

File: cmd/utils/flags.go (hunk 4 above)
Comment (jeongkyun-oh) [reply cmd/utils/flags.go:603]: As discussed in f2f, it is helpful while debugging.

File: cmd/utils/flags.go (hunk 12 above)
Comment (jeongkyun-oh) [reply cmd/utils/flags.go]: I merged two flags into one.

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 14 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: This was for debugging. Removed them 7e352ad7f4e399834adb85bad542131ccd7fa8a0.

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 13 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Right. Even `ChainDataFetcher` has the wait group variable, so no need to pass it as a parameter. Applied on a9ca7a8536ad08739a1a40fc515a807023667ece.

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+	defer f.wg.Done()
+	for {
+		select {
+		case <-f.stopCh:
+			logger.Info("stopped a handler")
+			return
+		case req := <-f.reqCh:
+			// TODO-ChainDataFetcher do handle new request
+			f.resCh <- req
+		}
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/chaindata_fetcher.go:100]: Is there any reason for receiving `req.Ch` instread of receiving `chainCh` directly?

File: datasync/chaindatafetcher/config.go [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+var DefaultChainDataFetcherConfig = &ChainDataFetcherConfig{
+	EnabledChainDataFetcher: false,
+	NoDefaultStart:          false,
+	NumHandlers:             10,
+	JobChannelSize:          50,
+
+	BlockChannelSize: 500,
+
+	DBPort: "3306",
+}
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/config.go:27]: How about using these values for initial flag values in `cmd/utils/flags.go`?

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+func (f *ChainDataFetcher) SetComponents(components []interface{}) {
+	for _, component := range components {
+		switch v := component.(type) {
+		case *blockchain.BlockChain:
+			f.blockchain = v
+		}
+	}
+}
+
+func (f *ChainDataFetcher) handler() {
```
Comment (aidan-kwon) [inline datasync/chaindatafetcher/chaindata_fetcher.go:89]: Please, consider renaming this method. `handler` sounds like returning handlers, but actually generates/adds a handler.

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 15 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go:100]: Yes, the request can be made not only from a chain event, but also from API.

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 17 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Applied. c7bb50563296063b05fd4c487934232bb06aa0ec

File: datasync/chaindatafetcher/config.go (hunk 16 above)
Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/config.go]: Applied. 5cd2c1301d87de8d6fe5a90a7d959cb13915867a

File: datasync/chaindatafetcher/chaindata_fetcher.go (hunk 15 above)
Comment (aidan-kwon) [reply datasync/chaindatafetcher/chaindata_fetcher.go:100]: I think `handleRequest` can handle the request from API if `f.chainSub` subscribes another channel receiving events from API. Then, we can reduce the number of loops. 
Uhm... Don't mind too much. It's not a big deal.

File: datasync/chaindatafetcher/chaindata_fetcher.go [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+	}
+}
+
+func (f *ChainDataFetcher) resLoop() {
+	f.wg.Add(1)
+	defer f.wg.Done()
+	for {
+		select {
+		case <-f.stopCh:
+			logger.Info("stopped reqLoop for chaindatafetcher")
```
Comment (ehnuje) [inline datasync/chaindatafetcher/chaindata_fetcher.go:125]:
```suggestion
			logger.Info("stopped resLoop for chaindatafetcher")
```

Comment (jeongkyun-oh) [reply datasync/chaindatafetcher/chaindata_fetcher.go]: Applied 1080eab4366732396fbc21354de189a8725120f0

**Pull request merged**